src/rm_cms.c

```diff
@@ -0,0 +1,328 @@
+    RedisModule_AutoMemory(ctx);
+
+    if(argc < 4) {
+        return RedisModule_WrongArity(ctx);
+    }   
+
+    long long numKeys = 0;
+    CMSketch *dest = NULL;   
+    CMSketch **cmsArray = CMS_CALLOC(numKeys, sizeof(CMSketch *));
+    long long *weights = CMS_CALLOC(numKeys, sizeof(long long));
+    if(parseMergeArgs(ctx, argv, argc, &dest, cmsArray, weights, numKeys) != 
+                            REDISMODULE_OK) {
+        CMS_FREE(cmsArray);
+        CMS_FREE(weights);
+        return REDISMODULE_OK;
+    }
+    
+    CMS_Merge(dest, numKeys, cmsArray, weights);
+
+    CMS_FREE(cmsArray);
+    CMS_FREE(weights);
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    return REDISMODULE_OK;
+}
+
+int CMSKetch_info(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+    
+    if (argc != 2) return RedisModule_WrongArity(ctx);
+
+    CMSketch *cms = NULL;    
+    if(GetCMSKey(ctx, argv[1], &cms, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_OK;
+    }
+
+    RedisModule_ReplyWithArray(ctx, 3*2);
+    RedisModule_ReplyWithSimpleString(ctx, "width");
+    RedisModule_ReplyWithLongLong(ctx, cms->width);
+    RedisModule_ReplyWithSimpleString(ctx, "depth");
+    RedisModule_ReplyWithLongLong(ctx, cms->depth);
+    RedisModule_ReplyWithSimpleString(ctx, "count");
+    RedisModule_ReplyWithLongLong(ctx, cms->counter);
+ 
+    return REDISMODULE_OK;
+}
+
+void CMSRdbSave(RedisModuleIO *io, void *obj) {
+    CMSketch *cms = obj;
+    RedisModule_SaveUnsigned(io, cms->width);
+    RedisModule_SaveUnsigned(io, cms->depth);
+    RedisModule_SaveUnsigned(io, cms->counter);
+    RedisModule_SaveStringBuffer(io, (const char *)cms->array, 
+                                cms->width * cms->depth * sizeof(size_t));
+}
+
+void *CMSRdbLoad(RedisModuleIO *io, int encver) {
+    if(encver > CMS_ENC_VER) {
+        return NULL;
+    }
+
+    CMSketch *cms = CMS_CALLOC(1, sizeof(CMSketch));
+    cms->width = RedisModule_LoadUnsigned(io);
+    cms->depth = RedisModule_LoadUnsigned(io);
+    cms->counter = RedisModule_LoadUnsigned(io);
+    size_t length = cms->width * cms->depth * sizeof(size_t);
+    cms->array = (size_t *)RedisModule_LoadStringBuffer(io, &length);
+
+    return cms;
+}
+
+void CMSFree(void *value) { 
+    CMS_Destroy(value);
+}
+
+size_t CMSMemUsage(const void *value) {
+    CMSketch *cms = (CMSketch *)value;
+    return sizeof(cms) + cms->width * cms->depth * sizeof(size_t);
+}
+
+int CMSModule_onLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModuleTypeMethods tm = {
+            .version = REDISMODULE_TYPE_METHOD_VERSION,
+            .rdb_load = CMSRdbLoad,
+            .rdb_save = CMSRdbSave,
+            .aof_rewrite = RMUtil_DefaultAofRewrite, 
+            .mem_usage = CMSMemUsage,
+            .free = CMSFree
+        };
+
+    CMSketchType = RedisModule_CreateDataType(ctx, "CMSk-TYPE", CMS_ENC_VER, &tm);
```

Comment:
I believe there are some naming restrictions on the module itself, I'd just use `cms`

Reply:
It must be 9 chars

Reply:
IIRC @ashtul is correct